shm/Player.hpp

```diff
@@ -0,0 +1,61 @@
+#pragma once
+#include <memory>
+
+#include "Ship.hpp"
+
+class Player {
+	std::unique_ptr<Ship> ship_;
+	int money_;
+	size_t availableSpace_;
+
+public:
+	Player():
+		ship_(std::make_unique<Ship>()),
+		money_(0),
+		availableSpace_()
+	{}
+
+	int getMoney() const { return money_; }
+
+	size_t getAvailableSpace() const { return availableSpace_; }
+
+	size_t getSpeed() const { 
+		if (ship_) {
+			return ship_->getSpeed();
+		}
+		return 0;
+	}
+
+	Cargo* getCargo(size_t index) const {
+		if (ship_) {
+			return ship_->getCargo(index);
+		}
+		return nullptr;
+	}
+
+	std::string takeOverShip(std::unique_ptr<Ship> someoneElsesShip) { 
+		if (!someoneElsesShip) {
+			return "there is no ship";
+		}
+		if (ship_) {
+			return "already have one";
+		}
+		ship_ = std::move(someoneElsesShip);
+		return "yarr";
+	}
+
+	std::unique_ptr<Ship> giveAwayShip() { 
+		return std::move(ship_); 
+	}
+
+	int calculateAvailableSpace() const {
+		size_t index = 0;
+		size_t count = 0;
+		auto cargo = getCargo(index);		
+		while (cargo) {
+			count += cargo->getAmount();
+			cargo = getCargo(++index);
+		}
+		return static_cast<int>(getAvailableSpace() - count);
```

Comment:
You can use std::accumulate.

Maybe we should cache this value? Notice, that this is an expensive function to calculate available space. And calculated value is valid until we add/remove cargo. So it is worth caching it. 
Think about `std::pair<bool, size_t>` where bool is `true` when value is valid and `false` when invalid. So you can immediately return `size_t` when a flag is `true` and do recalculation only when is `false`. 
There are also other ways, try to think how you want to avoid unnecessary calculations :)